### ISSUE TITLE
Relative system mode.

### ISSUE THREAD

**Issue by chronokun** [ApoorvaJ/EasyTab#9]
Took me a while and some reading through the wacom sdk docs to figure out how to make it work in relative mode but in the end it only took these 3 lines of code added to EasyTab_Load()

`LogContext.lcSysMode = 1;
LogContext.lcSysSensX = 5000;
LogContext.lcSysSensY = 5000;`

However it would be nice if some information about this were added to the usage examples along with how to disable system mode if you don't want the tablet moving the cursor etc.

Cheers for the handy library btw =)

**Comment by serge-rgb** [comment]
Kudos on doing the research. The whole point of this library is to get around the Wacom docs and API. It looks to me like this is a feature that this API should have.
Perhaps a `#define`?

```
#define EASYTAB_RELATIVE_MODE
#define EASYTAB_IMPLEMENTATION
#include "easytab.h"
```

and it would Just Work, setting things up to work in relative mode, which seems to be three lines on Windows. 

Linux support would be required. I would be happy to do it sometime towards the end of the month, when I will have lots of free time.

Doing it at runtime with a flag in the EasyTab struct is an option, and not doing it at all is another option. 

What does Apoorva think? Shall we do it? Shan't? :)

**Comment by ApoorvaJ** [comment]
Thanks for your interest in the lib and for the research, @chronokun. Diving through the dense Wacom docs is a chore and you did the dirty work for us. :)

Added the code for this in above pull request. 

@serge-rgb 
Please review. 
- Do you think a bool (e.g. IsRelativeMode) would be better than an enum as a parameter in the EasyTab_Load() function?
- I decided to keep it as a param to EasyTab_Load() instead of a struct var because we can't change this at runtime (in our lib, at least), and I don't really like the "populate struct, then call function" pattern instead of just calling function with params, especially when number of params is small.

Any suggestions/comments/disagreements welcome. I'll merge after you've reviewed and approved.

**Comment by serge-rgb** [comment]
I think It's a good idea to use an enum. Clearer and future-proof.

This approach breaks C99 compatibility, because of default parameters. I think C compatibility is important, and not only as a C user :)

Regardless, `Sensitivity` either needs to be more verbose (`RelativeModeSentivity` ?) or it needs documentation so the user knows that it only works for `mode == EASYTAB_MODE_RELATIVE`. 

Since there are no default parameters, how about increasing the surface area? Have a single EasyTab_Load() that does the obvious thing like it always did, and a EasyTab_Load_Ex(...) (with a better name..) with parameters and/or a struct. I would prefer a struct var. The benefit of a struct IMO is that it is future proof. Whenever the API needs extending, you can grow the struct and the calling code remains unchanged until it needs the new functionality.

**Comment by chronokun** [comment]
I think it would also be good to have to option to set whether or not to set the `CXO_SYSTEM` bit in `lcOptions` to enable/disable the system mouse cursor being linked to the tablet input (I'm not sure what a good name for that is but nearly anything would be more descriptive than 'system cursor context'). With it disabled the sensitivity needs to be set by `lcSensX` and `lcSensY` instead of `lcSysSensX` and `lcSysSensX` . I'm also not sure what the unit for sensitivity actually is, 5000 just happened to feel about right for my preferences.

Also I agree that it would be better to keep things C compatible.

**Comment by serge-rgb** [comment]
@chronokun Would it be too much to ask for a proof of concept pull request? If not, a github gist would be very helpful too. It seems to me like at this moment you are the one who has this stuff fresh. There is no time pressure at all.

**Comment by chronokun** [comment]
Sure, I guess so =)

Edit: Ok, I've done that now. #11

**Comment by serge-rgb** [comment]
Thanks for taking the time @chronokun. 
There are a couple of issues, but Apoorva is fixing them as I write this.
- Should be tabs, not spaces.
- Including stdbool.h in the "header" section of the library would break a lot of code, since it defines `bool`, `true`, and `false`, if they haven't been defined already.

Other than that. It seems like this does the job! :) Thanks again.

**Comment by ApoorvaJ** [comment]
Thanks for your pull request @chronokun, but there were a few issues with your pull request such as tabs instead of spaces, a new data structure, which I think might be overkill, and bools.

So I've updated my own pull request for now. We'll have to add good documentation to indicated that MouseMove is treated like a boolean value, and that if a user elects to use relative mode, 5000 (and not 0) is a good value for sensitivity. Maybe drop an `assert(RelativeModeSensitivity != 0)` in there with appropriate comment.

Let me know what you think, and whether I've made any mistakes. I don't have a Wacom tablet on me at the moment, and so I'm kinda winging it right now.

**Comment by serge-rgb** [comment]
Without benevolent dictators, shit wouldn't get done. 

I disagree with the solution for the reasons I have already written, but the important thing is that shit got done. We should not spend a lot of time on 20 lines of code that we can easily change later. If we need a struct var, we will add it when it's necessary.  :)

(edit: I should say that I noticed that the system cursor changes are not in apoorva's PR)

**Comment by ApoorvaJ** [comment]
I missed the system cursor code while diffing by eye. I have fixed that, and pushed to master.

We can change the param list to a var later, if you feel strongly about it.

Hopefully the code should work now. As said earlier, I don't have a Wacom tablet on me for the next few days, so this is pretty much a blind change, and I've only ensured that it compiles correctly on my machine.

I will close this pull request tomorrow, giving you guys some more time to maybe run the code and spot any mistakes I might have made.

**Comment by chronokun** [comment]
Yeah, I wasn't sure about using bools since they're built-in with C++ and not in C, was thinking about using a typedef or enum instead, but your solution using an int32_t works so I don't mind =)

Edit: I think there are some more issues that we both missed, eg. `ScreenToClient(Window, &Point);` is still being called in `EasyTab_HandleEvent()` when relative input is enabled, I'll send a new pull request later after I've tested more, going to try and get buttons being handled in a reasonable way too since I need that for what I'm working on atm.

Edit2: Also need to figure out how to detect if the pen has been lifted out of communication range to correct for the large position change when it starts generating events again.

Edit3: 

> FIX32
> A 32-bit fixed-point arithmetic type, with the radix point between the two words. Thus, the type contains 16 bits to the left of the radix point and 16 bits to the right of it.

So 0x00010000 is 1.0 sensitivity?

Edit4: New pull request #13

**Comment by ApoorvaJ** [comment]
Closed #13, and commented there. Please have a look. Closing this issue too.
